client/src/frontend/components/content-block/Block/types/TOCBlock/Contents.js

```diff
@@ -0,0 +1,151 @@
+import React, { PureComponent } from "react";
+import PropTypes from "prop-types";
+import { Link } from "react-router-dom";
+import classNames from "classnames";
+import lh from "helpers/linkHandler";
+
+export default class TocBlockToc extends PureComponent {
+  static displayName = "TocBlock.Toc";
+
+  static propTypes = {
+    showTextTitle: PropTypes.bool,
+    showAuthors: PropTypes.bool,
+    text: PropTypes.object.isRequired,
+    depth: PropTypes.number,
+    blockClass: PropTypes.string
+  };
+
+  static defaultProps = {
+    blockClass: "toc-block"
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.keyCounter = 0;
+  }
+
+  get textTitle() {
+    return this.props.text.attributes.titleFormatted;
+  }
+
+  get textSubtitle() {
+    return this.props.text.attributes.subtitle;
+  }
+
+  get toc() {
+    return this.props.text.attributes.toc;
+  }
+
+  showChildren = (node, depth) => {
+    return (
+      depth < this.props.depth && node.children && node.children.length > 0
+    );
+  };
+
+  renderNode = (node, depth = 1) => {
+    const blockClass = this.props.blockClass;
+
+    let children = null;
+
+    if (this.showChildren(node, depth)) {
+      depth++; // eslint-disable-line no-param-reassign
+
+      children = (
+        <ul
+          className={classNames(
+            `${blockClass}__list`,
+            `${blockClass}__list--depth-${depth}`
+          )}
+        >
+          {/* eslint-disable-next-line no-shadow */}
+          {node.children.map(node => this.renderNode(node, depth))}
+        </ul>
+      );
+    }
+
+    const anchor = node.anchor ? `#${node.anchor}` : "";
+
+    return (
+      <li key={this.keyCounter++} className={`${blockClass}__node`}>
```

Comment:
Just a suggestion: you could eliminate the need for this entire counter by using the node's id instead.

Reply:
I was getting duplicate keys initially, but maybe after fixing how the depth was being tracked, this would work. Thanks.

Reply:
@SMaxOwok I switched this to `node.id` but am getting the duplicate keys warning. Is it clear to you what I'm doing wrong?

Reply:
Hmm, I'll pull and take a quick look.  I dunno why they wouldn't be unique.

Reply:
Ok, so it's because sometimes a line in the TOC is to a specific part of a page.  There can be more than one link to the same page.

`${node.id}${anchor}` works reliably, but I dunno if it's necessarily better than the counter you had set up.

Curious if @zdavis has any feelings, although this thread has probably turned into a useless suggestion... sorry!

Reply:
It's still possible that a TOC could include two links that were identical. In this case, because we can be pretty darn certain the JSX won't change after it's rendered, I'm OK with using an index-like counter, even though Eslint warns against it. That said, a better solution would be to include a UUID for each TOC entry on the server-side. That would involve writing a migration, of course, and updating some ingestion code.